cardano-cli/src/Cardano/CLI/Ping.hs

```diff
@@ -0,0 +1,188 @@
+import           Text.Show (Show(..))
+
+import qualified Cardano.Network.Ping as CNP
+import qualified Control.Concurrent.Class.MonadSTM.Strict as STM
+import qualified Data.List as L
+import qualified Data.Text as T
+import qualified Network.Socket as Socket
+import qualified Options.Applicative as Opt
+import qualified Prettyprinter as PP
+import qualified System.Exit as IO
+import qualified System.IO as IO
+
+data PingClientCmdError
+  = PingClientCmdErrorOfInvalidHostIp
+  | PingClientCmdErrorOfExceptions ![(AddrInfo, SomeException)]
+
+
+data PingCmd = PingCmd
+  { pingCmdCount    :: !Word32
+  , pingCmdHost     :: !(Maybe String)
+  , pingCmdUnixSock :: !(Maybe String)
+  , pingCmdPort     :: !String
+  , pingCmdMagic    :: !Word32
+  , pingCmdJson     :: !Bool
+  , pingCmdQuiet    :: !Bool
+  } deriving (Eq, Show)
+
+pingClient :: Tracer IO CNP.LogMsg -> Tracer IO String -> PingCmd -> [CNP.NodeVersion] -> AddrInfo -> IO ()
+pingClient stdout stderr cmd versions peer =
+  CNP.pingClient stdout stderr opts versions peer
+  where opts = CNP.PingOpts
+          { CNP.pingOptsQuiet     = pingCmdQuiet cmd
+          , CNP.pingOptsJson      = pingCmdJson cmd
+          , CNP.pingOptsCount     = pingCmdCount cmd
+          , CNP.pingOptsHost      = pingCmdHost cmd
+          , CNP.pingOptsUnixSock  = pingCmdUnixSock cmd
+          , CNP.pingOptsPort      = pingCmdPort cmd
+          , CNP.pingOptsMagic     = pingCmdMagic cmd
+          }
+
+runPingCmd :: PingCmd -> ExceptT PingClientCmdError IO ()
+runPingCmd options = do
+  let hints = Socket.defaultHints { Socket.addrSocketType = Socket.Stream }
+
+  msgQueue <- liftIO STM.newEmptyTMVarIO
+
+  when (isNothing (pingCmdHost options) && isNothing (pingCmdUnixSock options)) $
+    throwE PingClientCmdErrorOfInvalidHostIp
+
+  (addresses, versions) <- case pingCmdUnixSock options of
+    Nothing -> do
+      addrs <- liftIO $ Socket.getAddrInfo (Just hints) (pingCmdHost options) (Just (pingCmdPort options))
+      return (addrs, CNP.supportedNodeToNodeVersions $ pingCmdMagic options)
+    Just fname ->
+      return
+        ( [ Socket.AddrInfo [] Socket.AF_UNIX Socket.Stream
+              Socket.defaultProtocol (Socket.SockAddrUnix fname)
+              Nothing
+          ]
+        , CNP.supportedNodeToClientVersions $ pingCmdMagic options
+        )
+
+  laid <- liftIO . async $ CNP.logger msgQueue $ pingCmdJson options
+  caids <- forM addresses $ liftIO . async . pingClient (Tracer $ doLog msgQueue) (Tracer doErrLog) options versions
+  res <- L.zip addresses <$> mapM (liftIO . waitCatch) caids
+  liftIO $ doLog msgQueue CNP.LogEnd
+  liftIO $ wait laid
+  case foldl' partition ([],[]) res of
+    ([], _) -> liftIO IO.exitSuccess
+    (es, []) -> throwE $ PingClientCmdErrorOfExceptions es
+    (es, _) -> do
+      unless (pingCmdQuiet options) $ mapM_ (liftIO . IO.hPrint IO.stderr) es
+      liftIO IO.exitSuccess
+
+  where
+    partition :: ([(AddrInfo, SomeException)], [AddrInfo])
```

Comment:
Some comments here would be useful

Reply:
Do the comments in the function body suffice?